lib/controlplane/agent/agent.c

```diff
@@ -966,3 +973,27 @@ void
 yanet_counter_handle_list_free(struct counter_handle_list *counters) {
 	free(counters);
 }
+
+void
+agent_free_unused_modules(struct agent *agent) {
+	while (agent->unused_module != NULL) {
+		struct cp_module *cp_module = ADDR_OF(&agent->unused_module);
+		SET_OFFSET_OF(&agent->unused_module, ADDR_OF(&cp_module->prev));
+		cp_module->free_handler(cp_module);
+	}
+}
+
+void
+agent_free_unused_agents(struct agent *agent) {
+	while (ADDR_OF(&agent->prev) != NULL) {
+		struct agent *prev_agent = ADDR_OF(&agent->prev);
+
+		if (prev_agent->loaded_module_count == 0) {
+			SET_OFFSET_OF(&agent->prev, ADDR_OF(&prev_agent->prev));
+			agent_cleanup(prev_agent);
+			continue;
+		}
+
+		agent = ADDR_OF(&agent->prev);
```

Comment:
It can lead to a null pointer dereference during the next check of the while condition. The loop conditions need to be reworked.